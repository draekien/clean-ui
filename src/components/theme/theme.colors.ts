export const colors = {
  text: '#292929',
  background: '#F7F7F7',
  primary: '#EF802F',
  secondary: '#DB5192',
  muted: '#D8D8D8',

  'p-000': '#FCE6D5',
  'p-100': '#F9CCAC',
  'p-200': '#F5B382',
  'p-300': '#F29959',
  'p-400': '#EF802F',

  's-000': '#F8DCE9',
  's-100': '#F1B9D3',
  's-200': '#E997BE',
  's-300': '#E274A8',
  's-400': '#DB5192',

  'b-000': '#F7F7F7',
  'b-100': '#EFEFEF',
  'b-200': '#D8D8D8',
  'b-300': '#B1B1B1',
  'b-400': '#898989',
  'b-500': '#626262',
  'b-600': '#3B3B3B',
  'b-700': '#292929',

  'info-000': '#CDE1EB',
  'info-400': '#066899',

  'success-000': '#D5EADC',
  'success-400': '#2E9852',

  'warning-000': '#FBDBD1',
  'warning-400': '#EA4318',

  'error-000': '#F9CCD4',
  'error-400': '#E00027',

  modes: {
    dark: {
      text: '#F7F7F7',
      background: '#292929',
      primary: '#EF802F',
      secondary: '#DB5192',
      muted: '#3B3B3B',
    },
  },
};
