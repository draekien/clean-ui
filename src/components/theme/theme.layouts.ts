export const layouts = {
  radii: {
    xs: '0.125rem',
    sm: '0.25rem',
    md: '0.5rem',
    lg: '0.75rem',
    xl: '1rem',
    max: 999,
  },
  space: {
    xxs: '0.25rem',
    xs: '0.5rem',
    sm: '0.75rem',
    md: '1rem',
    lg: '1.5rem',
    xl: '2rem',
    xxl: '2.5rem',
  },
  shadows: {
    sm: '0px 4px 4px rgba(61, 61, 61, 0.25), 0px 2px 4px -2px rgba(61, 61, 61, 0.25);',
    md: '0px 4px 8px rgba(61, 61, 61, 0.25), 0px 4px 8px -4px rgba(61, 61, 61, 0.25);',
    lg: '0px 4px 8px rgba(61, 61, 61, 0.25), 0px 16px 32px -16px rgba(61, 61, 61, 0.25)',
    text: '0px 2px 4px rgba(61, 61, 61, 0.25);',
    disabled: '0px 2px 4px -2px rgba(4, 40, 52, 0.25);',
  },
  zIndices: {
    lowest: 0,
    lower: 100,
    low: 200,
    middle: 300,
    high: 400,
    higher: 500,
    highest: 600,
    pagination: 1,
    topMenu: 200,
    dropdown: 300,
    drawer: 400,
    modal: 500,
    toast: 600,
    overlay: 600,
  },
};
