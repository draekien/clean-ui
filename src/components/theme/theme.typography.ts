export const typography = {
  fonts: {
    body: '"Barlow", sans-serif',
    mono: '"Fira Code", monospace',
  },
  fontSizes: {
    hero: '6rem',
    h1: '3rem',
    h2: '2.5rem',
    h3: '2rem',
    h4: '1.5rem',
    subtitle: '1rem',
    body: '1rem',
    eyebrow: '0.75rem',
    small: '0.75rem',
  },
  fontWeights: {
    thin: 300,
    regular: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
  },
  lineHeights: {
    body: 1.5,
    heading: 1.125,
  },
};
