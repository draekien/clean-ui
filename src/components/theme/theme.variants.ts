import { ThemeUIStyleObject } from 'theme-ui';

export const textVariants: Record<string, ThemeUIStyleObject> = {
  hero: {
    fontSize: 'hero',
    fontWeight: 'bold',
    lineHeight: 'heading',
  },
  h1: {
    fontSize: 'h1',
    fontWeight: 'bold',
    lineHeight: 'heading',
  },
  h2: {
    fontSize: 'h2',
    fontWeight: 'bold',
    lineHeight: 'heading',
  },
  h3: {
    fontSize: 'h3',
    fontWeight: 'bold',
    lineHeight: 'heading',
  },
  h4: {
    fontSize: 'h4',
    fontWeight: 'bold',
    lineHeight: 'heading',
  },
  subtitle: {
    fontSize: 'subtitle',
    fontWeight: 'semibold',
    lineHeight: 'heading',
  },
  body: {
    fontSize: 'body',
    fontWeight: 'regular',
    lineHeight: 'body',
  },
  eyebrow: {
    fontSize: 'eyebrow',
    fontWeight: 'medium',
    lineHeight: 'body',
    textTransform: 'uppercase',
  },
  small: {
    fontSize: 'small',
    fontWeight: 'regular',
    lineHeight: 'body',
  },
  mono: {
    fontFamily: 'mono',
  },
};

export const variants = {
  text: textVariants,
};
