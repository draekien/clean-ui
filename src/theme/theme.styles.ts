import { ThemeStyles } from '@theme-ui/css';

export const styles: ThemeStyles = {
  root: {
    fontFamily: 'body',
    fontWeight: 'body',
    fontSize: '16px',
    '*:focus': {
      outline: 'none',
    },
  },
};
