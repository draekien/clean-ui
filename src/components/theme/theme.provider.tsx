/** @jsxImportSource theme-ui */
import React, { useEffect } from 'react';
import { ThemeProvider as ThemeUiProvider } from 'theme-ui';
import { CleanUiTheme, Theme } from './theme';

export interface CleanUiProps {
  theme?: Partial<Theme>;
  children: React.ReactNode;
}

const mergeThemes = (target: any, source: any) => {
  Object.keys(source).forEach((key) => {
    if (key === '__proto__' || key === 'constructor') return;

    if (source[key] && typeof source[key] === 'object') {
      mergeThemes((target[key] = target[key] || {}), source[key]);
      return;
    }

    target[key] = source[key];
  });
};

const addThemeFonts = () => {
  const fonts: { [key: string]: string } = {
    text:
      'https://fonts.googleapis.com/css2?family=Barlow:wght@200;300;400;500;600;700&family=Fira+Code&display=swap',
    icons:
      'https://fonts.googleapis.com/icon?family=Material+Icons|Material+Icons+Outlined|Material+Icons+Two+Tone|Material+Icons+Round|Material+Icons+Sharp',
  };

  const preconnectLink = document.createElement('link');
  preconnectLink.rel = 'preconnect';
  preconnectLink.href = 'https://fonts.gstatic.com';
  document.head.appendChild(preconnectLink);

  Object.keys(fonts).forEach((key) => {
    const linkExists = document.getElementById(key);

    if (!linkExists) {
      const link = document.createElement('link');
      link.id = key;
      link.rel = 'stylesheet';
      link.href = fonts[key];
      link.crossOrigin = 'anonymous';
      document.head.appendChild(link);
    }
  });
};

export const CleanUi = ({ theme, children }: CleanUiProps) => {
  const newTheme = JSON.parse(JSON.stringify(CleanUiTheme));

  if (theme) mergeThemes(newTheme, theme);

  useEffect(() => {
    addThemeFonts();
  }, []);

  return <ThemeUiProvider theme={newTheme}>{children}</ThemeUiProvider>;
};
