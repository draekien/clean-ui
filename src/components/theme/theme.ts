import { Theme as ThemeUiTheme } from '@theme-ui/css';
import { colors } from './theme.colors';
import { layouts } from './theme.layouts';
import { styles } from './theme.styles';
import { typography } from './theme.typography';
import { variants } from './theme.variants';

export type Theme = Partial<{}>;

export const CleanUiTheme: ThemeUiTheme = {
  useCustomProperties: false,
  useBorderBox: true,
  useRootStyles: true,
  useLocalStorage: true,
  breakpoints: ['40em', '56em', '64em'],
  colors,
  styles,
  ...variants,
  ...layouts,
  ...typography,
};
