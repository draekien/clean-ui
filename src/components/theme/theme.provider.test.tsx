import { render } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from './theme.provider';

describe('Theme Provider', () => {
  it('Should merge themes', () => {
    const theme = {
      space: { xl: '5rem' },
      __proto__: 'something bad',
      constructor: 'something worse',
    };
    const { container } = render(
      <ThemeProvider theme={theme}>
        <div />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
