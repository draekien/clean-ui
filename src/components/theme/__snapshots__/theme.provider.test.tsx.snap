// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme Provider Should merge themes 1`] = `
<div>
  <div />
</div>
`;
